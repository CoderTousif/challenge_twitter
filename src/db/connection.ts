import Knex from 'knex'
import KnexTinyLogger from 'knex-tiny-logger'

const config = require('../../knexfile')[process.env.NODE_ENV || 'development']

const knex =
  process.env.NODE_ENV === 'development'
    ? KnexTinyLogger(Knex(config))
    : Knex(config)

export default knex
