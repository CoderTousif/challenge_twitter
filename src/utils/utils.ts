import User from '../entities/User'
import jwt from 'jsonwebtoken'
import { JWT_SECRET } from '../config/config'
import { GraphQLFormattedError } from 'graphql'

export const generateToken = (user: User) => {
  const token = jwt.sign(
    {
      data: {
        id: user.id,
        username: user.username,
        display_name: user.display_name,
      },
    },
    JWT_SECRET as string,
    { expiresIn: '7d' } // 7 days
  )
  return token
}
