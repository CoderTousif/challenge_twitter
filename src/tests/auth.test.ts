import { ValidationError } from 'apollo-server'
import db from '../db/connection'
import { generateToken } from '../utils/utils'
import { createUser } from './helpers'
import { LOGIN, ME, REGISTER } from './queries/auth.queries'
import { testClient } from './setup'

describe('Auth', () => {
  beforeEach(async () => {
    await db.migrate.rollback()
    await db.migrate.latest()
  })

  afterEach(async () => {
    await db.migrate.rollback()
  })

  test('it should register a user', async () => {
    const { mutate } = await testClient()

    const res = await mutate({
      mutation: REGISTER,
      variables: {
        input: {
          username: 'admin',
          display_name: 'Admin',
          email: 'admin@test.fr',
          password: 'password',
        },
      },
    })

    const [newUser] = await db('users').where('username', 'admin')

    expect(newUser).not.toBeUndefined
    expect(newUser.username).toEqual('admin')
  })

  test('it should not register a user if the username is incorrect', async () => {
    const { mutate } = await testClient()

    const res = await mutate({
      mutation: REGISTER,
      variables: {
        input: {
          username: 'admin)',
          display_name: 'Admin',
          email: 'admin@test.fr',
          password: 'password',
        },
      },
    })
    expect(res).toMatchSnapshot()
    expect(res.data).toBeNull()
  })

  test('it should not register a user if the email already exists', async () => {
    await createUser('admin', 'admin@test.fr')

    const { mutate } = await testClient()

    const res = await mutate({
      mutation: REGISTER,
      variables: {
        input: {
          username: 'new',
          display_name: 'Admin',
          email: 'admin@test.fr',
          password: 'password',
        },
      },
    })

    expect(res.errors).not.toBeNull()

    const {
      extensions: {
        exception: { validationErrors },
      },
    }: any = res.errors![0]

    expect((validationErrors[0] as ValidationError).constraints).toEqual({
      UniqueConstraint: 'This email is already taken',
    })
    expect(res.data).toBeNull()
  })

  test('should log in a user', async () => {
    await createUser()
    const { mutate } = await testClient()

    const res = await mutate({
      mutation: LOGIN,
      variables: {
        input: {
          email: 'admin@test.fr',
          password: 'password',
        },
      },
    })

    const { token, user } = res.data.login
    expect(token).not.toBeNull()
    expect(user.username).toEqual('admin')
    expect(user.email).toEqual('admin@test.fr')
  })

  test('should throw a validation error if the email is invalid', async () => {
    await createUser()
    const { mutate } = await testClient()

    const res = await mutate({
      mutation: LOGIN,
      variables: {
        input: {
          email: 'adminaz',
          password: 'password',
        },
      },
    })

    expect(res.data).toBeNull()
    expect(res.errors).not.toBeNull()

    expect(res.errors![0].extensions).toMatchSnapshot()
    const {
      extensions: {
        exception: { validationErrors },
      },
    }: any = res.errors![0]

    expect((validationErrors[0] as ValidationError).constraints).toEqual({
      isEmail: 'email must be an email',
    })
  })

  test('should throw a validation error if the password is empty', async () => {
    await createUser()
    const { mutate } = await testClient()

    const res = await mutate({
      mutation: LOGIN,
      variables: {
        input: {
          email: 'admin@test.fr',
          password: '',
        },
      },
    })

    expect(res.data).toBeNull()
    expect(res.errors).not.toBeNull()
    expect(res.errors![0].extensions).toMatchSnapshot()
    const {
      extensions: {
        exception: { validationErrors },
      },
    }: any = res.errors![0]

    expect((validationErrors[0] as ValidationError).constraints).toEqual({
      isNotEmpty: 'password should not be empty',
    })
  })

  test('it should fetch the currentUser', async () => {
    const user = await createUser()

    const { query } = await testClient({
      req: { headers: { authorization: 'Bearer ' + generateToken(user) } },
    })

    const res = await query({
      query: ME,
    })

    expect(res.data).not.toBeNull()
    expect(res.data.me.id).toEqual(user.id)
  })

  test('it should throw an unauthorized error if there is no token', async () => {
    const user = await createUser()

    const { query } = await testClient()

    const res = await query({
      query: ME,
    })

    expect(res).toMatchSnapshot()
  })

  test('it should throw expired Token error', async () => {
    const user = await createUser()

    const { query } = await testClient({
      req: { headers: { authorization: 'Bearer ' + generateToken(user, -60) } },
    })

    const res = await query({
      query: ME,
    })

    expect(res).toMatchSnapshot()
  })
})
