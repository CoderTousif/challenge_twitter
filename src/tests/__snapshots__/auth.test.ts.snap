// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should not register a user if the username is incorrect 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Argument Validation Error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`it should throw an unauthorized error if there is no token 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Unauthorized],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`it should throw expired Token error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: jwt expired],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`should throw a validation error if the email is invalid 1`] = `
Object {
  "code": "INTERNAL_SERVER_ERROR",
  "exception": Object {
    "validationErrors": Array [
      ValidationError {
        "children": Array [],
        "constraints": Object {
          "isEmail": "email must be an email",
        },
        "property": "email",
        "target": LoginPayload {
          "email": "adminaz",
          "password": "password",
        },
        "value": "adminaz",
      },
    ],
  },
}
`;

exports[`should throw a validation error if the password is empty 1`] = `
Object {
  "code": "INTERNAL_SERVER_ERROR",
  "exception": Object {
    "validationErrors": Array [
      ValidationError {
        "children": Array [],
        "constraints": Object {
          "isNotEmpty": "password should not be empty",
        },
        "property": "password",
        "target": LoginPayload {
          "email": "admin@test.fr",
          "password": "",
        },
        "value": "",
      },
    ],
  },
}
`;
